### DISCUSSION TITLE
Support TypedDict NotRequired

### DISCUSSION THREAD

**Discussion by Kludex** [pydantic/pydantic#3563] [Question] [answered]
This is a question/feature request. Will Pydantic support NotRequired?

References: 
- https://mypy-lang.blogspot.com/2021/12/mypy-0930-released.html?spref=tw
- https://www.python.org/dev/peps/pep-0655/

Snippet that raises error:
```python
from pydantic import BaseModel
from typing_extensions import NotRequired, TypedDict


class Potato(TypedDict):
    color: NotRequired[str]

class Model(BaseModel):
    potato: Potato

print(Model(potato=Potato()))
```

Sorry for the short presentation, I'm tired. I think it's enough, but I can add more details later on if asked/needed.

**Comment by eastside** [comment]
I'm also a little puzzled by how to represent, in pydantic, a Dictionary that has a field that could be missing, or could be a particular type.

Here's how you'd do this with standard Python type hinting.

```py
from typing_extensions import TypedDict, NotRequired

class Movie(TypedDict):
    title: str
    year: NotRequired[int]

Movie(title='The Matrix', year=1999)    # Validation pass
Movie(title='The Matrix')               # Validation pass
Movie(title='The Matrix', year='asdf')  # Validation failure: year should be int
Movie(title='The Matrix', year=None)    # Validation failure: year should not be None
```

But in Pydantic, I'm not sure how you would achieve this. With Optional[] you could get case 1, 2, and 3, but not 4.

I'm very new to Pydantic so would love a hint if I'm unaware of some feature.

**Comment by microHoffman** [comment]
Bumping this as I am wondering the same:).

**Reply by samuelcolvin** [reply]
This is fixed properly in pydantic-core and will work as expected in V2.

It won't get fixed before then.

**Reply by microHoffman** [reply]
Nice, thanks for the info:)

**Comment by clbarnes** [comment]
Will it only be in effect for TypedDicts in pydantic 2, or could it be used in BaseModels and dataclasses, effectively adding to [this table](https://docs.pydantic.dev/2.0/migration/#required-optional-and-nullable-fields)?

**Reply by Kludex** [reply]
Only `TypedDict`, since it's a _thing_ only for `TypedDict`.

**Reply by clbarnes** [reply]
I suppose it doesn't matter too much because python doesn't have distinct `null` and `undefined` types (although I suppose a singleton could be introduced for the latter), so undefined fields would only be represented by a `None` anyway. It would be nice to distinguish between `{}` and `{"field": null}` though, as some specs say "`field` MAY exist, in which case it has type X"; you don't necessarily want to serialise that to a `null`.

**Comment by Kludex** [comment, marked as answer]
This was fixed in https://github.com/pydantic/pydantic/pull/3687. 🙏
